code/game/machinery/doors/firedoor.dm

```diff
@@ -88,11 +94,236 @@
 		return
 	if(operating || !density)
 		return
-	user.changeNext_move(CLICK_CD_MELEE)
+	if(!isliving(user))
+		return
+	var/mob/living/living_user = user
+	if(living_user.combat_mode)
+		living_user.changeNext_move(CLICK_CD_MELEE)
+		living_user.visible_message("<span class='notice'>[living_user] bangs on [src].</span>", \
+			"<span class='notice'>You bang on [src].</span>")
+		playsound(loc, 'sound/effects/glassknock.ogg', 10, FALSE, frequency = 32000)
+	else
+		if(DOING_INTERACTION(living_user, DOAFTER_SOURCE_BAREHAND_FIRELOCK_OPEN))//you get one firelock to open, better pick the right one
+			return
+
+		//dont want them to be moved by pressure differences when theyre holding onto the firelock with all their strength
+		living_user.move_resist = MOVE_FORCE_VERY_STRONG
+
+		//starting stats in order to write logs that provide enough data to make kickass fucking graphs later
+		var/starting_health = living_user.health
+		var/starting_time = REALTIMEOFDAY
+
+		var/datum/gas_mixture/users_air = living_user.return_air()
+		var/user_loc_starting_temperature = users_air.temperature
+
+		///used so that we can log everyone's stats once when the do_after finishes
+		var/list/user_stat_list = list(list("user" = living_user, "starting temperature" = user_loc_starting_temperature, "starting health" = starting_health, "starting time" = starting_time))
```

Comment:
Is this assoc list necessary? Is it used for logging? We should be avoiding assoc lists for things like this for memory reasons.

Reply:
The alternative, if this is not used for readable logs, is to make it a normal numerically indexed array with every index being a define, as in, instead of `list("user" = living_user, "starting temperature" = starting_temperature)`, you'd do `list(living_user, starting_temperature)`, and then:

```dm
#define KEY_LIVING_USER 1
#define KEY_STARTING_TEMPERATURE 2
```